File: Algorithm/Programers_대충 만든 자판.cpp
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;

vector<int> solution(vector<string> keymap, vector<string> targets) {
	vector<int> answer;
	unordered_map<char, int> key;
	for (auto iter : keymap) {
		for (int i = 0; i < iter.size(); i++) {
			if (key.find(iter[i]) == key.end()) {
				key[iter[i]] = i + 1;
				continue;
			}
			if (key[iter[i]] > i + 1)
				key[iter[i]] = i + 1;
		}
	}
	for (auto iter : targets) {
		int keyCnt = 0;
		for (int i = 0; i < iter.size(); i++) {
			if (key.find(iter[i]) == key.end()) {
				keyCnt = -1;
				break;
			}
			keyCnt += key[iter[i]];
		}
		answer.push_back(keyCnt);
	}

	return answer;
}
//https://school.programmers.co.kr/learn/courses/30/lessons/160586

File: Algorithm/Programers_대충_만든_자판 .java
import java.util.*;

import java.util.*;
class Solution {
    public int[] solution(String[] keymap, String[] targets) {
        int[] answer = new int[targets.length];
        HashMap<Character,Integer> map = new HashMap<>();

        for (String key : keymap) {
            for (int i = 0; i < key.length(); i++) {
                char ch = key.charAt(i);
                if (!map.containsKey(ch)){
                    map.put(ch,i+1);
                    continue;
                }
                if(map.get(ch)>i){
                    map.replace(ch,i+1);
                }

            }
        }

        for(int i=0; i< targets.length;i++){
            int sum = 0;
            for(int j=0; j<targets[i].length();j++){

                char ch = targets[i].charAt(j);

                if(!map.containsKey(ch)){
                    sum = -1;
                    break;
                }
                sum+=map.get(ch);
            }
            answer[i] = sum;
        }

        return answer;
    }
}
class main{
    public static void main(String[] args) {
        Solution sol = new Solution();
        String keymap[] = {"ABACD", "BCEFD"};
        String targets[] = {"ABCD","AABB"};
        System.out.println("sol = " + sol.solution(keymap,targets));
    }
}
//https://school.programmers.co.kr/learn/courses/30/lessons/160586